#### Data description:
Aruba:
- 1920-1938
- Daily Pressure & Rainfall

ArubaS:
- 1936-1940
- Daily measurements of various variables measured (sometimes whole table, sometimes only wind)

Bonaire:
- 1920-1941
- Daily rain, max and min pressure and temperature (**not always**)

Curacao:
- 1898-1903
- Hourly wind speed

CuracaoM:
- 1902 - 1919
- Various subdaily

Saba:
- 1920 - 1945
- Subdaily pressure, min and max temperature and daily rain

Statia:
- 1943 - 1946
- Subdaily pressure, daily rain

Willemstad:
- 1912 - 1946
- Various subdaily measurements
- TXT format

#### Notes:
- Lot of variability in time scale (hourly, daily, subdaily)
- Lot of variability in recorded variables (between 1 and 12 variables) -> sparse data if all data is put in one table with all columns


- Values with time vs without time? -> 12 uur voor metingen zonder tijd. Of Tijd kolom die Null is. 
- Lege kolommen? -> mnu
- Suriname kolommen?
- Null vs NaN vs None

In [1]:
import duckdb
import pandas as pd 
import os
from tqdm.auto import tqdm

#### Convert to CSV

In [43]:
# importing pandas as pd 
# read an excel file and convert  
# into a dataframe object 
dirs = ['Aruba', 'Bonaire', 'Curacao', 'Saba', 'Statia']
for dir in tqdm(dirs):
    if not os.path.exists(f'CSVs/{dir}'):
        os.makedirs(f'CSVs/{dir}')
    for file in os.listdir(f'digitized/{dir}'):
        for sheet in ['Regen', 'Overig']:
            try:
                name, ext = file.split('.')
                df = pd.DataFrame(pd.read_excel(f"digitized/{dir}/{file}", sheet_name=sheet)) 
                df.to_csv(f'CSVs/{dir}/{name}_{sheet}.csv')
            except ValueError:
                # Catch sheet not found error from files without all sheets
                print(f"Error processing {file} in {sheet} sheet")
                continue

  0%|          | 0/5 [00:00<?, ?it/s]

Error processing Curacao1901.xlsm in Regen sheet
Error processing Curacao1899.xlsm in Regen sheet
Error processing Curacao1903.xlsm in Regen sheet
Error processing Curacao1902.xlsm in Regen sheet
Error processing Curacao1898.xlsm in Regen sheet
Error processing Curacao1900.xlsm in Regen sheet


In [ ]:
# Special treatment for Suriname (xls files, and no separate sheets for rain and other 
# resulting in a two level header)
if not os.path.exists('CSVs/Suriname'):
    os.makedirs('CSVs/Suriname')
for file in os.listdir('digitized/Suriname'):
    name, ext = file.split('.')
    df = pd.DataFrame(pd.read_excel(f"digitized/Suriname/{file}", header=[0,1])) 
    df.to_csv(f'CSVs/Suriname/{name}.csv')

In [ ]:
# Special treatment for Willemstad (latin1 encoded TXT files following CSV format)
import csv
if not os.path.exists('CSVs/Willemstad'):
    os.makedirs('CSVs/Willemstad')
for file in os.listdir('digitized/Willemstad'):
    name, ext = file.split('.')
    df = pd.read_csv(f"digitized/Willemstad/{file}", delimiter=',', encoding='latin1')
    # use QUOTE_NONE to remove double quotes from the TXT file
    df.to_csv(f'CSVs/Willemstad/{name}.csv', quoting=csv.QUOTE_NONE) 

#### DuckDB Config

In [5]:
%load_ext sql
conn = duckdb.connect('knmi_database.db')
%sql conn --alias duckdb

%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

#### Aruba

In [111]:
%%sql
DROP TABLE IF EXISTS Aruba;
CREATE TABLE IF NOT EXISTS Aruba AS
SELECT * FROM
(SELECT * FROM read_csv('CSVs/Aruba/*Regen.csv', union_by_name = true, dateformat = '%Y%m%d'))
FULL JOIN 
(SELECT * FROM read_csv('CSVs/Aruba/*Overig.csv', union_by_name = true, dateformat = '%Y%m%d'))
USING (Stn, Datum)
;
SELECT * FROM Aruba LIMIT 1;

column0  Stn      Datum  Niveau Rd  qRd  column0_1      P  qP  column00  \
0        0   32 1920-01-01       1  0    0          0  764.0   0      <NA>   

   ...    Ed   qEd    Tw   qTw     D    qD     F    qF     N    qN  
0  ...  None  <NA>  None  <NA>  None  <NA>  None  <NA>  None  <NA>  

[1 rows x 29 columns]

In [124]:
cursor = conn.execute("PRAGMA table_info('Aruba');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Aruba
    SET {column} = NULL
    WHERE TRIM({column}) = '—';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column0', 'Stn', 'Datum', 'Niveau', 'qRd', 'column0_1', 'qP', 'column00', 'Tijd', 'qT', 'qTn', 'qTx', 'qU', 'qEd', 'qTw', 'qD', 'qF', 'qN']


In [167]:
%%sql
df << SELECT column_name, column_type, min, max, approx_unique, count, null_percentage
FROM (SUMMARIZE Aruba)

In [168]:
df.to_latex('Aruba_summary.tex', index=False, float_format="%.2f", escape=False, column_format='lcccccc')

In [126]:
%%sql
SELECT *, datetime AT TIME ZONE 'America/Aruba' AS timestamp FROM
(SELECT *, Datum + to_hours(COALESCE(CAST(Tijd AS INTEGER),24)) AS Datetime FROM Aruba)
LIMIT 1; 

column0  Stn      Datum  Niveau Rd  qRd  column0_1      P  qP  column00  \
0        0   32 1920-01-01       1  0    0          0  764.0   0      <NA>   

   ...    Tw   qTw     D    qD     F    qF     N    qN   Datetime  \
0  ...  None  <NA>  None  <NA>  None  <NA>  None  <NA> 1920-01-02   

                  timestamp  
0 1920-01-02 04:00:00+00:00  

[1 rows x 31 columns]

#### Bonaire

In [8]:
%%sql
CREATE TABLE IF NOT EXISTS Bonaire AS
(SELECT *  FROM
  (SELECT * FROM read_csv('CSVs/Bonaire/*Regen.csv', union_by_name = true, dateformat = '%Y%m%d')) t1
FULL JOIN 
  (SELECT * FROM read_csv('CSVs/Bonaire/*Overig.csv', union_by_name = true, dateformat = '%Y%m%d')) t2
USING (Stn, Datum))
;

Empty DataFrame
Columns: [Count]
Index: []

In [127]:
cursor = conn.execute("PRAGMA table_info('Bonaire');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Bonaire
    SET {column} = NULL
    WHERE TRIM({column}) = '—';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column0', 'Stn', 'Datum', 'Niveau', 'qRd', 'column00', 'Tijd', 'Px', 'qPx', 'Pn', 'qPn', 'Tx', 'qTx', 'Tn', 'qTn', 'column0_1', 'qP', 'qT']


In [9]:
%%sql
SELECT *, datetime AT TIME ZONE 'America/Aruba' AS timestamp FROM
(SELECT *, Datum + to_hours(COALESCE(CAST(Tijd AS INTEGER),24)) AS Datetime FROM Bonaire)
LIMIT 1;

column0  Stn      Datum  Niveau Rd  qRd  column00  Tijd     Px  qPx  ...  \
0        0   44 1920-01-01       1  0    0         0  24.0  757.0    0  ...   

   qTx    Tn  qTn  column0_1     P    qP     T    qT   Datetime  \
0    0  28.5    0       <NA>  None  <NA>  None  <NA> 1920-01-02   

                  timestamp  
0 1920-01-02 04:00:00+00:00  

[1 rows x 23 columns]

In [128]:
%sql FROM (SUMMARIZE Bonaire);

column_name column_type         min         max  approx_unique  \
0      column0      BIGINT           0        1097           1278   
1          Stn      BIGINT          42          49              3   
2        Datum        DATE  1920-01-01  1941-12-31           5459   
3       Niveau      BIGINT           1           9             10   
4           Rd     VARCHAR           0          99            291   
5          qRd      BIGINT           0           7              2   
6     column00      BIGINT           0         365            385   
7         Tijd      DOUBLE         8.0        24.0              3   
8           Px      DOUBLE       753.0       760.5             14   
9          qPx      BIGINT           0           0              1   
10          Pn      DOUBLE       751.5       759.0             16   
11         qPn      BIGINT           0           0              1   
12          Tx      DOUBLE        26.0        31.5             12   
13         qTx      BIGINT           0           0              1   
14          Tn      DOUBLE        25.0        31.0             13   
15         qTn      BIGINT           0           0              1   
16   column0_1      BIGINT           0         731            840   
17           P     VARCHAR     733.933     838.454            390   
18          qP      BIGINT           0           7              2   
19           T     VARCHAR        20.5        33.0             43   
20          qT      BIGINT           0           7              2   

                   avg                 std                 q25  \
0   394.66771191080585  287.04041368132647                 162   
1    44.14533870887462  2.1860411835670304                  42   
2                 None                None          1930-12-01   
3    1.143945050022398  1.0614684765908877                   1   
4                 None                None                None   
5   0.4501518092678314   1.717140872020826                   0   
6                182.5  105.79933837222235                  91   
7    11.90400918403391   4.091752787139371                 8.0   
8    757.0765027322404  1.2201032108792575   756.3888888888889   
9                  0.0                 0.0                   0   
10    755.389344262295   1.341914847984338               754.5   
11                 0.0                 0.0                   0   
12  29.101092896174862   0.951940430451125                28.5   
13                 0.0                 0.0                   0   
14  28.165300546448087  1.0530442717148596  27.333333333333332   
15                 0.0                 0.0                   0   
16  364.76720204416864  210.89716203302396                 182   
17                None                None                None   
18  0.2734075561233802   1.356196537477642                   0   
19                None                None                None   
20  0.2734075561233802   1.356196537477642                   0   

                   q50         q75  count  null_percentage  
0                  324         601  20091             0.00  
1                   44          44  20091             0.00  
2           1935-02-15  1938-07-26  20091             0.00  
3                    1           1  20091             0.00  
4                 None        None  20091             6.43  
5                    0           0  20091             0.00  
6                  182         274  20091            98.18  
7   14.976716315699367        15.0  20091            43.64  
8                757.0       758.0  20091            98.18  
9                    0           0  20091            98.18  
10              755.09  756.265625  20091            98.18  
11                   0           0  20091            98.18  
12                29.0        30.0  20091            98.18  
13                   0           0  20091            98.18  
14                28.0        29.0  20091            98.18  
15                   0           0  20091            98

#### Curacao

In [11]:
%%sql
CREATE TABLE IF NOT EXISTS Curacao AS
SELECT * FROM
  (SELECT * FROM read_csv('CSVs/Curacao/*Regen.csv', union_by_name = true, dateformat = '%Y%m%d')) t1
FULL JOIN 
  (SELECT * FROM read_csv('CSVs/Curacao/*Overig.csv', union_by_name = true, dateformat = '%Y%m%d')) t2
USING (Stn, Datum)
;

Empty DataFrame
Columns: [Count]
Index: []

In [129]:
cursor = conn.execute("PRAGMA table_info('Curacao');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Curacao
    SET {column} = NULL
    WHERE TRIM({column}) = '—';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column0', 'Stn', 'Datum', 'Niveau', 'qRd', 'column0_1', 'Tijd', 'qF', 'column00', 'qTp', 'qPt', 'qPc', 'qP0', 'qP', 'qT', 'qTw', 'qTd', 'qU', 'qEd', 'qTx', 'qTn', 'qD', 'qFx', 'qDFx', 'qN', 'qC', 'qDn']


In [130]:
%%sql
SELECT *, datetime AT TIME ZONE 'America/Aruba' AS timestamp FROM
(SELECT *, Datum + to_hours(COALESCE(CAST(Tijd AS INTEGER),24)) AS Datetime FROM Curacao)
LIMIT 1;


column0    Stn      Datum  Niveau    Rd  qRd  column0_1   Tijd     F  qF  \
0        0  78989 1907-01-01       1  None    7       <NA>  800.0  None   7   

   ...   DFx qDFx     N qN     C qC    Dn qDn            Datetime  \
0  ...  None    7  None  7  None  7  None   7 1907-02-03 08:00:00   

                  timestamp  
0 1907-02-03 12:00:00+00:00  

[1 rows x 49 columns]

In [131]:
%sql FROM (SUMMARIZE Curacao);

column_name column_type         min         max  approx_unique  \
0      column0      BIGINT           0         365            385   
1          Stn      BIGINT       78989       78989              1   
2        Datum        DATE  1898-01-01  1919-12-31           7321   
3       Niveau      BIGINT           1           1              1   
4           Rd     VARCHAR           0           T            119   
5          qRd      BIGINT           0           7              2   
6    column0_1      BIGINT           0        8759           7828   
7         Tijd      DOUBLE         1.0       800.0             32   
8            F     VARCHAR           0         9.8             41   
9           qF      BIGINT           0           7              3   
10    column00      BIGINT           0         365            385   
11          Tp     VARCHAR        23.9        31.4             32   
12         qTp      BIGINT           0           7              2   
13          Pt     VARCHAR      757.53       766.8            225   
14         qPt      BIGINT           0           7              2   
15          Pc     VARCHAR        6.27        7.04             30   
16         qPc      BIGINT           0           7              2   
17          P0     VARCHAR      747.04      761.21            267   
18         qP0      BIGINT           0           7              2   
19           P     VARCHAR       753.1       763.8             38   
20          qP      BIGINT           0           7              2   
21           T     VARCHAR        None        None              0   
22          qT      BIGINT           7           7              1   
23          Tw     VARCHAR        None        None              0   
24         qTw      BIGINT           7           7              1   
25          Td     VARCHAR       -12.7       -13.1              5   
26         qTd      BIGINT           0           7              2   
27           U     VARCHAR          73          81             10   
28          qU      BIGINT           0           7              2   
29          Ed     VARCHAR        18.9        20.2             10   
30         qEd      BIGINT           0           7              2   
31          Tx     VARCHAR        25.2        36.2             87   
32         qTx      BIGINT           0           7              2   
33          Tn     VARCHAR         1.9        36.1            108   
34         qTn      BIGINT           0           7              2   
35           D     VARCHAR           0          90             10   
36          qD      BIGINT           0           7              2   
37          Fx     VARCHAR        None        None              0   
38         qFx      BIGINT           7           7              1   
39         DFx     VARCHAR        None        None              0   
40        qDFx      BIGINT           7           7              1   
41           N     VARCHAR        None        None              0   
42          qN      BIGINT           7           7              1   
43           C     VARCHAR        None        None              0   
44          qC      BIGINT           7           7              1   
45          Dn     VARCHAR           0          90              5   
46         qDn      BIGINT           0           7              2   

                   avg                 std                q25  \
0   182.02673581942364  105.38346926636754                 91   
1              78989.0                 0.0              78989   
2                 None                None         1899-10-13   
3                  1.0                 0.0                  1   
4                 None                None               None   
5    4.871288213594361   3.220240269106866                  0   
6               4379.5  2528.8182191779947               2191   
7   126.70981592143418   265.6118209818324  7.771365407785331   
8                 None                None               None   
9   2.2830548390714367   3.2812384057053

#### Saba

In [13]:
%%sql
DROP TABLE IF EXISTS Saba;
CREATE TABLE IF NOT EXISTS Saba AS
(SELECT *  FROM
  (SELECT * FROM read_csv('CSVs/Saba/*Regen.csv', union_by_name = true, dateformat = '%Y%m%d')) t1
FULL JOIN 
  (SELECT * FROM read_csv('CSVs/Saba/*Overig.csv', union_by_name = true, dateformat = '%Y%m%d')) t2
USING (Stn, Datum))
;

Count
0  38726

In [134]:
cursor = conn.execute("PRAGMA table_info('Saba');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Saba
    SET {column} = NULL
    WHERE TRIM({column}) = '—';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column0', 'Stn', 'Datum', 'Niveau', 'qRd', 'column0_1', 'Tijd', 'qP', 'qTn', 'qTx', 'qT']


In [132]:
%%sql
SELECT *, datetime AT TIME ZONE 'America/Aruba' AS timestamp FROM
(SELECT *, Datum + to_hours(COALESCE(CAST(Tijd AS INTEGER),24)) AS Datetime FROM Saba)
LIMIT 1;

column0  Stn      Datum  Niveau  Rd  qRd  column0_1  Tijd    P  qP Tn  qTn  \
0        0   81 1920-01-01       1  48    0          0   8.0  749   0  —    7   

  Tx  qTx     T    qT            Datetime                 timestamp  
0  —    7  None  <NA> 1920-01-01 08:00:00 1920-01-01 12:00:00+00:00

In [135]:
%sql FROM (SUMMARIZE Saba);

column_name column_type         min         max  approx_unique  \
0      column0      BIGINT           0        1097           1278   
1          Stn      BIGINT          81          83              3   
2        Datum        DATE  1920-01-01  1945-12-31           7080   
3       Niveau      BIGINT           1           1              1   
4           Rd     VARCHAR           0          99            370   
5          qRd      BIGINT           0           7              3   
6    column0_1      BIGINT           0        1097           1278   
7         Tijd      DOUBLE        7.45        24.0              5   
8            P     VARCHAR         458       762.8            236   
9           qP      BIGINT           0           7              2   
10          Tn     VARCHAR        22.3        29.3             34   
11         qTn      BIGINT           0           7              2   
12          Tx     VARCHAR        22.5          32             31   
13         qTx      BIGINT           0           7              2   
14           T     VARCHAR          20          49             36   
15          qT      BIGINT           0           7              2   

                   avg                 std         q25         q50  \
0   388.92160305737747  306.99926774092404         146         293   
1     81.5659763466405  0.7892913106056159          81          81   
2                 None                None  1929-02-09  1934-05-24   
3                  1.0                 0.0           1           1   
4                 None                None        None        None   
5   1.4528223932241904   2.838558206621826           0           0   
6    547.5464045782533  316.42226637890207         274         548   
7   13.535755162982028  5.0045606534180855        7.45       13.45   
8                 None                None        None        None   
9   1.5466533963672555    2.90427044781216           0           0   
10                None                None        None        None   
11   4.666666666666667   3.301335328297214           0           7   
12                None                None        None        None   
13   4.666666666666667   3.301335328297214           0           7   
14                None                None        None        None   
15  1.5018248175182483  2.8736135298600534           0           0   

           q75  count  null_percentage  
0          609  38726             0.00  
1           82  38726             0.00  
2   1939-09-12  38726             0.00  
3            1  38726             0.00  
4         None  38726            20.75  
5            0  38726             0.00  
6          822  38726            37.73  
7        19.45  38726            37.73  
8         None  38726            51.49  
9            0  38726            37.73  
10        None  38726            99.05  
11           7  38726            97.16  
12        None  38726            99.05  
13           7  38726            97.16  
14        None  38726            53.32  
15           0  38726            40.57

#### Statia

In [15]:
%%sql
DROP TABLE Statia;
CREATE TABLE IF NOT EXISTS Statia AS
(SELECT *  FROM
  (SELECT * FROM read_csv('CSVs/Statia/*Regen.csv', union_by_name = true, dateformat = '%Y%m%d')) t1
FULL JOIN 
  (SELECT * FROM read_csv('CSVs/Statia/*Overig.csv', union_by_name = true, dateformat = '%Y%m%d')) t2
USING (Stn, Datum))
;

Count
0   4383

In [138]:
cursor = conn.execute("PRAGMA table_info('Statia');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Statia
    SET {column} = NULL
    WHERE TRIM({column}) = '—';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column0', 'Stn', 'Datum', 'Niveau', 'qRd', 'column0_1', 'Tijd', 'qP']


In [136]:
%%sql
SELECT *, datetime AT TIME ZONE 'America/Aruba' AS timestamp FROM
(SELECT *, Datum + to_hours(COALESCE(CAST(Tijd AS INTEGER),24)) AS Datetime FROM Statia)
LIMIT 1;

column0  Stn      Datum  Niveau Rd  qRd  column0_1  Tijd    P  qP  \
0        0   73 1943-01-01       1  0    0          2  16.0  761   0   

             Datetime                 timestamp  
0 1943-01-01 16:00:00 1943-01-01 20:00:00+01:00

In [139]:
%sql FROM (SUMMARIZE Statia);

column_name column_type         min         max  approx_unique  \
0     column0      BIGINT           0         365            385   
1         Stn      BIGINT          73          73              1   
2       Datum        DATE  1943-01-01  1946-12-31           1732   
3      Niveau      BIGINT           1           2              2   
4          Rd     VARCHAR           0          98            169   
5         qRd      BIGINT           0           7              2   
6   column0_1      BIGINT           0        1097           1278   
7        Tijd      DOUBLE         8.0        16.0              3   
8           P     VARCHAR         755         766             13   
9          qP      BIGINT           0           7              2   

                   avg                   std         q25         q50  \
0    182.1252566735113     105.4506729078786          91         182   
1                 73.0                   0.0          73          73   
2                 None                  None  1944-01-01  1944-12-31   
3    1.000684462696783  0.026156266923742105           1           1   
4                 None                  None        None        None   
5  0.29705681040383297    1.4112438904436446           0           0   
6    547.3757700205339    316.35307264087663         274         547   
7                 12.0    3.2663589617774624         8.0        12.0   
8                 None                  None        None        None   
9   3.9383983572895276    3.4728314885410096           0           7   

          q75  count  null_percentage  
0         273   4383             0.00  
1          73   4383             0.00  
2  1945-12-31   4383             0.00  
3           1   4383             0.00  
4        None   4383             4.24  
5           0   4383             0.00  
6         822   4383             0.00  
7        16.0   4383             0.00  
8        None   4383            56.26  
9           7   4383             0.00

#### Suriname

In [17]:
%%sql 
DROP TABLE IF EXISTS Suriname;
CREATE TABLE IF NOT EXISTS Suriname AS
SELECT
    column00 AS "column00",
    column01 AS "Datum",
    column02 AS "Luchtdruk_8",
    column03 AS "Luchtdruk_12",
    column04 AS "Luchtdruk_18",
    column05 AS "Temp_droog_8",
    column06 AS "Temp_nat_8",
    column07 AS "Temp_verschil_8",
    column08 AS "Temp_droog_12",
    column09 AS "Temp_nat_12",
    column10 AS "Temp_verschil_12",
    column11 AS "Temp_droog_18",
    column12 AS "Temp_nat_18",
    column13 AS "Temp_verschil_18",
    column14 AS "Temp_min",
    column15 AS "Temp_max",
    column16 AS "Dampdruk_8",
    column17 AS "Dampdruk_12",
    column18 AS "Dampdruk_18",
    column19 AS "Rel_vocht_8",
    column20 AS "Rel_vocht_12",
    column21 AS "Rel_vocht_18",
    column22 AS "Wind_richt_8",
    column23 AS "Wind_kracht_8",
    column24 AS "Wind_richt_12",
    column25 AS "Wind_kracht_12",
    column26 AS "Wind_richt_18",
    column27 AS "Wind_kracht_18",
    column28 AS "Bewolking_8",
    column29 AS "Bewolking_12",
    column30 AS "Bewolking_18",
    column31 AS "Regen",
    column32 AS "Regen_wrong",
FROM read_csv('CSVs/Suriname/*.csv', union_by_name = true, dateformat = '%Y%m%d', skip = 2, null_padding = true, ignore_errors = true);

Count
0  25275

In [149]:
cursor = conn.execute("PRAGMA table_info('Suriname');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Suriname
    SET {column} = NULL
    WHERE TRIM({column}) = '-';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column00', 'Datum', 'Luchtdruk_8', 'Luchtdruk_12', 'Luchtdruk_18', 'Temp_droog_8', 'Temp_verschil_8', 'Wind_kracht_12', 'Regen']


In [150]:
%%sql
FROM (SUMMARIZE Suriname)

column_name column_type                 min                 max  \
0           column00      BIGINT                   0                 377   
1              Datum      DOUBLE          18960101.0          19621231.0   
2        Luchtdruk_8      DOUBLE                64.0  1017.2645161290322   
3       Luchtdruk_12      DOUBLE                10.0  1015.7193548387097   
4       Luchtdruk_18      DOUBLE                13.0  1014.8129032258064   
5       Temp_droog_8      DOUBLE  23.480645161290322   1016.058064516129   
6         Temp_nat_8     VARCHAR             1010.47                99.0   
7    Temp_verschil_8      DOUBLE               -37.0  1013.8387096774194   
8      Temp_droog_12     VARCHAR               183.0               399.0   
9        Temp_nat_12     VARCHAR               166.0               358.0   
10  Temp_verschil_12     VARCHAR               -27.0                 9.0   
11     Temp_droog_18     VARCHAR               116.0               348.0   
12       Temp_nat_18     VARCHAR               145.0               302.0   
13  Temp_verschil_18     VARCHAR                 0.0                99.0   
14          Temp_min     VARCHAR               158.0                99.0   
15          Temp_max     VARCHAR               100.0                99.0   
16        Dampdruk_8     VARCHAR                Calm                 WSW   
17       Dampdruk_12     VARCHAR                   0                   W   
18       Dampdruk_18     VARCHAR                 0.0                 WSW   
19       Rel_vocht_8     VARCHAR                 0.0                 WSW   
20      Rel_vocht_12     VARCHAR                 0.0                 WSW   
21      Rel_vocht_18     VARCHAR                 0.0                 WSW   
22      Wind_richt_8     VARCHAR                 0.0                 WSW   
23     Wind_kracht_8     VARCHAR                 0.0                  SE   
24     Wind_richt_12     VARCHAR                 0.0                 WSW   
25    Wind_kracht_12      DOUBLE                 0.0              1482.0   
26     Wind_richt_18     VARCHAR                 0.0                 WNW   
27    Wind_kracht_18     VARCHAR                   0                  96   
28       Bewolking_8     VARCHAR                 0.0                97.0   
29      Bewolking_12     VARCHAR                 0.0                 9.0   
30      Bewolking_18     VARCHAR                 0.0    9.67741935483871   
31             Regen      DOUBLE                 0.0              2022.0   
32       Regen_wrong     VARCHAR                 0.0                96.0   

    approx_unique                 avg                 std                 q25  \
0             406  188.11964391691396  108.90165251974767                  94   
1           22883    19290672.9398063  193392.08591152562  19122889.844526347   
2             940   536.9089220109886  193.98589778492993    603.746400706313   
3             927   524.1612515936762  195.98543415819904   588.4832858919547   
4            1261   520.3054917768216  197.34742294744035   585.7055634865461   
5             722  246.79448448074461    50.9395194693209  244.99873335135837   
6             944                None                None                None   
7            1131  232.86592606244582   85.82453974093721  250.50164919875738   
8             794                None                None                None   
9             930                None                None                None   
10            994                None                None                None   
11           1016                None                None                None   
12           1025                None                None                None   
13            666                None                None                None   
14            809                None                None                None   
15            847                None                None                None   
16            306                None          

#### Willemstad

In [19]:
%%sql
DROP TABLE IF EXISTS Willemstad;
CREATE TABLE IF NOT EXISTS Willemstad AS
SELECT * FROM read_csv('CSVs/Willemstad/*.csv', union_by_name = true, 
dateformat = '%Y%m%d', skip=1,  columns=
{
    'column00': 'BIGINT',
    'Stn': 'BIGINT',
    'Datum': 'DATE',
    'Tijd': 'BIGINT',
    'DD': 'VARCHAR',
    'qDD': 'BIGINT',
    'FK': 'VARCHAR',
    'FF': 'VARCHAR',
    'FF1': 'VARCHAR',
    'FF2': 'VARCHAR',
    'qFF': 'BIGINT',
    'WW': 'VARCHAR',
    'qWW': 'BIGINT',
    'N': 'VARCHAR',
    'qN': 'BIGINT',
    'T': 'VARCHAR',
    'qT': 'BIGINT',
    'U': 'VARCHAR',
    'qU': 'BIGINT',
    'EE': 'VARCHAR',
    'qEE': 'BIGINT',
    'Tw': 'VARCHAR',
    'qTw': 'BIGINT',
    'P': 'VARCHAR',
    'Pt': 'VARCHAR',
    'Pt1': 'VARCHAR',
    'qPt': 'BIGINT',
    'Rd': 'VARCHAR',
    'Dr': 'VARCHAR',
    'qRd': 'BIGINT',
    'EV': 'VARCHAR',
    'qEV': 'BIGINT',
    'Trn': 'VARCHAR',
    'Tin': 'VARCHAR',
    'Tn': 'VARCHAR',
    'qTn': 'BIGINT',
    'Trx': 'VARCHAR',
    'Tix': 'VARCHAR',
    'Tx': 'VARCHAR',
    'qTx': 'BIGINT'
});

Count
0  58803

In [ ]:
cursor = conn.execute("PRAGMA table_info('Willemstad');")
columns = [row[1] for row in cursor.fetchall()]
failed_columns = []

for column in columns:
    update_query = f"""
    UPDATE Willemstad
    SET {column} = NULL
    WHERE TRIM({column}) = '//////';
    """
    try:
        conn.execute(update_query)
    except Exception:
        failed_columns.append(column)
        continue
conn.commit()
print(f"Failed to update columns: {failed_columns}")

Failed to update columns: ['column00', 'Stn', 'Datum', 'Tijd', 'qDD', 'qFF', 'qWW', 'qN', 'qT', 'qU', 'qEE', 'qTw', 'qPt', 'qRd', 'qEV', 'qTn', 'qTx']


In [27]:
%%sql
UPDATE Willemstad
SET WW = NULL
WHERE TRIM(WW) = '//////////////';

Count
0  58803

In [28]:
%%sql
FROM (SUMMARIZE SELECT * FROM Willemstad);

column_name column_type                        min  \
0     column00      BIGINT                          0   
1          Stn      BIGINT                      78989   
2        Datum        DATE                 1910-01-01   
3         Tijd      BIGINT                        600   
4           DD     VARCHAR                          N   
5          qDD      BIGINT                          0   
6           FK     VARCHAR                          0   
7           FF     VARCHAR                          0   
8          FF1     VARCHAR                          0   
9          FF2     VARCHAR                          1   
10         qFF      BIGINT                          0   
11          WW     VARCHAR                       None   
12         qWW      BIGINT                          7   
13           N     VARCHAR                          0   
14          qN      BIGINT                          0   
15           T     VARCHAR                        195   
16          qT      BIGINT                          0   
17           U     VARCHAR                          0   
18          qU      BIGINT                          0   
19          EE     VARCHAR                          0   
20         qEE      BIGINT                          0   
21          Tw     VARCHAR                        185   
22         qTw      BIGINT                          0   
23           P     VARCHAR                       None   
24          Pt     VARCHAR                      72960   
25         Pt1     VARCHAR                       2972   
26         qPt      BIGINT                          0   
27          Rd     VARCHAR                          0   
28          Dr     VARCHAR   ////////////////////////   
29         qRd      BIGINT                          0   
30          EV     VARCHAR                       None   
31         qEV      BIGINT                          7   
32         Trn     VARCHAR                       None   
33         Tin     VARCHAR                       None   
34          Tn     VARCHAR                        170   
35         qTn      BIGINT                          0   
36         Trx     VARCHAR                       None   
37         Tix     VARCHAR                       None   
38          Tx     VARCHAR                        230   
39         qTx      BIGINT                          0   

                          max  approx_unique                 avg  \
0                        2919           2966   826.3362923660358   
1                       78989              1             78989.0   
2                  1946-12-31          14032                None   
3                        2400             13  1544.6790129755284   
4                         ZZW             16                None   
5                           7              2  2.2821454687686002   
6                          20             17                None   
7                          22             16                None   
8                          72             60                None   
9                          22             20                None   
10                          7              2   2.320357804873901   
11                       None              0                None   
12                          7              1                 7.0   
13                         16             10                None   
14                          7              2  2.2750029760386377   
15                        349            139                None   
16                          7              2   2.245361631209292   
17                        100             52                None   
18                          7              2  2.2454806727547916   
19                       4441            121                None   
20                          7              2  2.3036919885039877   
21                        294            104                None   
22                          7              2   2.249170960665272   
23                       None    

In [31]:
%%sql
FROM (SHOW ALL TABLES);

database schema        name  \
0  knmi_database   main       Aruba   
1  knmi_database   main     Bonaire   
2  knmi_database   main     Curacao   
3  knmi_database   main        Saba   
4  knmi_database   main      Statia   
5  knmi_database   main    Suriname   
6  knmi_database   main  Willemstad   

                                        column_names  \
0  [Stn, Datum, Niveau, Rd, qRd, P, qP, Tijd, T, ...   
1  [column0, Stn, Datum, Niveau, Rd, qRd, column0...   
2  [column0, Stn, Datum, Niveau, Rd, qRd, column0...   
3  [column0, Stn, Datum, Niveau, Rd, qRd, column0...   
4  [column0, Stn, Datum, Niveau, Rd, qRd, column0...   
5  [column00, Datum, Luchtdruk_8, Luchtdruk_12, L...   
6  [column00, Stn, Datum, Tijd, DD, qDD, FK, FF, ...   

                                        column_types  temporary  
0  [BIGINT, DATE, BIGINT, VARCHAR, BIGINT, VARCHA...      False  
1  [BIGINT, BIGINT, DATE, BIGINT, VARCHAR, BIGINT...      False  
2  [BIGINT, BIGINT, DATE, BIGINT, VARCHAR, BIGINT...      False  
3  [BIGINT, BIGINT, DATE, BIGINT, VARCHAR, BIGINT...      False  
4  [BIGINT, BIGINT, DATE, BIGINT, VARCHAR, BIGINT...      False  
5  [BIGINT, DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBL...      False  
6  [BIGINT, BIGINT, DATE, BIGINT, VARCHAR, BIGINT...      False

In [40]:
17537 + 20091 + 62419 + 38726 + 4383 + 25275 + 58803

227234

In [39]:
%%sql
FROM (SUMMARIZE Willemstad)

column_name column_type                        min  \
0     column00      BIGINT                          0   
1          Stn      BIGINT                      78989   
2        Datum        DATE                 1910-01-01   
3         Tijd      BIGINT                        600   
4           DD     VARCHAR                          N   
5          qDD      BIGINT                          0   
6           FK     VARCHAR                          0   
7           FF     VARCHAR                          0   
8          FF1     VARCHAR                          0   
9          FF2     VARCHAR                          1   
10         qFF      BIGINT                          0   
11          WW     VARCHAR                       None   
12         qWW      BIGINT                          7   
13           N     VARCHAR                          0   
14          qN      BIGINT                          0   
15           T     VARCHAR                        195   
16          qT      BIGINT                          0   
17           U     VARCHAR                          0   
18          qU      BIGINT                          0   
19          EE     VARCHAR                          0   
20         qEE      BIGINT                          0   
21          Tw     VARCHAR                        185   
22         qTw      BIGINT                          0   
23           P     VARCHAR                       None   
24          Pt     VARCHAR                      72960   
25         Pt1     VARCHAR                       2972   
26         qPt      BIGINT                          0   
27          Rd     VARCHAR                          0   
28          Dr     VARCHAR   ////////////////////////   
29         qRd      BIGINT                          0   
30          EV     VARCHAR                       None   
31         qEV      BIGINT                          7   
32         Trn     VARCHAR                       None   
33         Tin     VARCHAR                       None   
34          Tn     VARCHAR                        170   
35         qTn      BIGINT                          0   
36         Trx     VARCHAR                       None   
37         Tix     VARCHAR                       None   
38          Tx     VARCHAR                        230   
39         qTx      BIGINT                          0   

                          max  approx_unique                 avg  \
0                        2919           2966   826.3362923660358   
1                       78989              1             78989.0   
2                  1946-12-31          14032                None   
3                        2400             13  1544.6790129755284   
4                         ZZW             16                None   
5                           7              2  2.2821454687686002   
6                          20             17                None   
7                          22             16                None   
8                          72             60                None   
9                          22             20                None   
10                          7              2   2.320357804873901   
11                       None              0                None   
12                          7              1                 7.0   
13                         16             10                None   
14                          7              2  2.2750029760386377   
15                        349            139                None   
16                          7              2   2.245361631209292   
17                        100             52                None   
18                          7              2  2.2454806727547916   
19                       4441            121                None   
20                          7              2  2.3036919885039877   
21                        294            104                None   
22                          7              2   2.249170960665272   
23                       None    

#### Join tables

Check that weather stations don't overlap:
- Aruba: 32
- Bonaire: 42, 44, 49
- Curacao: 78989
- Saba: 83, 81, 82
- Statia: 73
- Suriname: 81201
- Willemstad: 78989

In [151]:
%%sql
FROM (SUMMARIZE SELECT * FROM Bonaire)

column_name column_type         min         max  approx_unique  \
0      column0      BIGINT           0        1097           1278   
1          Stn      BIGINT          42          49              3   
2        Datum        DATE  1920-01-01  1941-12-31           5459   
3       Niveau      BIGINT           1           9             10   
4           Rd     VARCHAR           0          99            291   
5          qRd      BIGINT           0           7              2   
6     column00      BIGINT           0         365            385   
7         Tijd      DOUBLE         8.0        24.0              3   
8           Px      DOUBLE       753.0       760.5             14   
9          qPx      BIGINT           0           0              1   
10          Pn      DOUBLE       751.5       759.0             16   
11         qPn      BIGINT           0           0              1   
12          Tx      DOUBLE        26.0        31.5             12   
13         qTx      BIGINT           0           0              1   
14          Tn      DOUBLE        25.0        31.0             13   
15         qTn      BIGINT           0           0              1   
16   column0_1      BIGINT           0         731            840   
17           P     VARCHAR     733.933     838.454            390   
18          qP      BIGINT           0           7              2   
19           T     VARCHAR        20.5        33.0             43   
20          qT      BIGINT           0           7              2   

                   avg                 std                 q25  \
0   394.66771191080585  287.04041368132647                 162   
1    44.14533870887462  2.1860411835670304                  42   
2                 None                None          1930-12-01   
3    1.143945050022398  1.0614684765908877                   1   
4                 None                None                None   
5   0.4501518092678314   1.717140872020826                   0   
6                182.5  105.79933837222235                  91   
7    11.90400918403391   4.091752787139371                 8.0   
8    757.0765027322404  1.2201032108792575   756.3888888888889   
9                  0.0                 0.0                   0   
10    755.389344262295   1.341914847984338               754.5   
11                 0.0                 0.0                   0   
12  29.101092896174862   0.951940430451125                28.5   
13                 0.0                 0.0                   0   
14  28.165300546448087  1.0530442717148596  27.333333333333332   
15                 0.0                 0.0                   0   
16  364.76720204416864  210.89716203302396                 182   
17                None                None                None   
18  0.2734075561233802   1.356196537477642                   0   
19                None                None                None   
20  0.2734075561233802   1.356196537477642                   0   

                   q50         q75  count  null_percentage  
0                  324         601  20091             0.00  
1                   44          44  20091             0.00  
2           1935-02-15  1938-07-26  20091             0.00  
3                    1           1  20091             0.00  
4                 None        None  20091             6.43  
5                    0           0  20091             0.00  
6                  182         274  20091            98.18  
7   14.976716315699367        15.0  20091            43.64  
8                757.0       758.0  20091            98.18  
9                    0           0  20091            98.18  
10              755.09  756.265625  20091            98.18  
11                   0           0  20091            98.18  
12                29.0        30.0  20091            98.18  
13                   0           0  20091            98.18  
14                28.0        29.0  20091            98.18  
15                   0           0  20091            98

In [90]:
%%sql
FROM (DESCRIBE SELECT * FROM Aruba)

column_name column_type null   key default extra
0          Stn      BIGINT  YES  None    None  None
1        Datum        DATE  YES  None    None  None
2       Niveau      BIGINT  YES  None    None  None
3           Rd     VARCHAR  YES  None    None  None
4          qRd      BIGINT  YES  None    None  None
5            P     VARCHAR  YES  None    None  None
6           qP      BIGINT  YES  None    None  None
7         Tijd      DOUBLE  YES  None    None  None
8            T     VARCHAR  YES  None    None  None
9           qT      BIGINT  YES  None    None  None
10          Tn     VARCHAR  YES  None    None  None
11         qTn      BIGINT  YES  None    None  None
12          Tx     VARCHAR  YES  None    None  None
13         qTx      BIGINT  YES  None    None  None
14           U     VARCHAR  YES  None    None  None
15          qU      BIGINT  YES  None    None  None
16          Ed     VARCHAR  YES  None    None  None
17         qEd      BIGINT  YES  None    None  None
18          Tw     VARCHAR  YES  None    None  None
19         qTw      BIGINT  YES  None    None  None
20           D     VARCHAR  YES  None    None  None
21          qD      BIGINT  YES  None    None  None
22           F     VARCHAR  YES  None    None  None
23          qF      BIGINT  YES  None    None  None
24           N     VARCHAR  YES  None    None  None
25          qN      BIGINT  YES  None    None  None

In [100]:
%%sql 
FROM (DESCRIBE Bonaire)

column_name column_type null   key default extra
0      column0      BIGINT  YES  None    None  None
1          Stn      BIGINT  YES  None    None  None
2        Datum        DATE  YES  None    None  None
3       Niveau      BIGINT  YES  None    None  None
4           Rd     VARCHAR  YES  None    None  None
5          qRd      BIGINT  YES  None    None  None
6     column00      BIGINT  YES  None    None  None
7         Tijd      DOUBLE  YES  None    None  None
8           Px      DOUBLE  YES  None    None  None
9          qPx      BIGINT  YES  None    None  None
10          Pn      DOUBLE  YES  None    None  None
11         qPn      BIGINT  YES  None    None  None
12          Tx      DOUBLE  YES  None    None  None
13         qTx      BIGINT  YES  None    None  None
14          Tn      DOUBLE  YES  None    None  None
15         qTn      BIGINT  YES  None    None  None
16   column0_1      BIGINT  YES  None    None  None
17           P     VARCHAR  YES  None    None  None
18          qP      BIGINT  YES  None    None  None
19           T     VARCHAR  YES  None    None  None
20          qT      BIGINT  YES  None    None  None

In [97]:
%%sql
FROM (DESCRIBE
SELECT 
    *
FROM
(SELECT * FROM Aruba) AS a
FULL JOIN
(SELECT * FROM Bonaire) AS b
USING (Stn, Datum))

column_name column_type null   key default extra
0          Stn      BIGINT  YES  None    None  None
1        Datum        DATE  YES  None    None  None
2       Niveau      BIGINT  YES  None    None  None
3           Rd     VARCHAR  YES  None    None  None
4          qRd      BIGINT  YES  None    None  None
5            P     VARCHAR  YES  None    None  None
6           qP      BIGINT  YES  None    None  None
7         Tijd      DOUBLE  YES  None    None  None
8            T     VARCHAR  YES  None    None  None
9           qT      BIGINT  YES  None    None  None
10          Tn     VARCHAR  YES  None    None  None
11         qTn      BIGINT  YES  None    None  None
12          Tx     VARCHAR  YES  None    None  None
13         qTx      BIGINT  YES  None    None  None
14           U     VARCHAR  YES  None    None  None
15          qU      BIGINT  YES  None    None  None
16          Ed     VARCHAR  YES  None    None  None
17         qEd      BIGINT  YES  None    None  None
18          Tw     VARCHAR  YES  None    None  None
19         qTw      BIGINT  YES  None    None  None
20           D     VARCHAR  YES  None    None  None
21          qD      BIGINT  YES  None    None  None
22           F     VARCHAR  YES  None    None  None
23          qF      BIGINT  YES  None    None  None
24           N     VARCHAR  YES  None    None  None
25          qN      BIGINT  YES  None    None  None
26     column0      BIGINT  YES  None    None  None
27      Niveau      BIGINT  YES  None    None  None
28          Rd     VARCHAR  YES  None    None  None
29         qRd      BIGINT  YES  None    None  None
30    column00      BIGINT  YES  None    None  None
31        Tijd      DOUBLE  YES  None    None  None
32          Px      DOUBLE  YES  None    None  None
33         qPx      BIGINT  YES  None    None  None
34          Pn      DOUBLE  YES  None    None  None
35         qPn      BIGINT  YES  None    None  None
36          Tx      DOUBLE  YES  None    None  None
37         qTx      BIGINT  YES  None    None  None
38          Tn      DOUBLE  YES  None    None  None
39         qTn      BIGINT  YES  None    None  None
40   column0_1      BIGINT  YES  None    None  None
41           P     VARCHAR  YES  None    None  None
42          qP      BIGINT  YES  None    None  None
43           T     VARCHAR  YES  None    None  None
44          qT      BIGINT  YES  None    None  None

In [162]:
%%sql
DROP TABLE IF EXISTS Aggregated;
CREATE TABLE Aggregated AS
SELECT 
    Stn,
    Datum,
    COALESCE (a.Niveau, b.Niveau) AS Niveau,
    COALESCE (a.Rd, b.Rd) AS Rd,
    COALESCE (a.qRd, b.qRd) AS qRd,
    COALESCE (a.Tijd, b.Tijd) AS Tijd,
    COALESCE (a.Tx, CAST(b.Tx AS VARCHAR)) AS Tx,
    COALESCE (a.qTx, b.qTx) AS qTx,
    COALESCE (a.Tn, CAST(b.Tn AS VARCHAR)) AS Tn,
    COALESCE (a.qTn, b.qTn) AS qTn,
    COALESCE (a.P, b.P) AS P,
    COALESCE (a.qP, b.qP) AS qP,
    COALESCE (a.T, b.T) AS T,
    COALESCE (a.qT, b.qT) AS qT,
    U, qU, Ed, qEd, Tw, qTw, D, qD, F, qF, N, qN,
    Px, qPx, Pn, qPn
FROM
(SELECT * FROM Aruba) AS a
FULL JOIN
(SELECT * FROM Bonaire) AS b
USING (Stn, Datum)

Count
0  37628

In [163]:
%%sql
FROM (SUMMARIZE Aggregated)

column_name column_type         min         max  approx_unique  \
0          Stn      BIGINT          32          49              4   
1        Datum        DATE  1920-01-01  1941-12-31           5459   
2       Niveau      BIGINT           1           9             10   
3           Rd     VARCHAR           0         990            345   
4          qRd      BIGINT           0           7              2   
5         Tijd      DOUBLE         7.0        24.0              6   
6           Tx     VARCHAR        26.0        34.6             70   
7          qTx      BIGINT           0           7              2   
8           Tn     VARCHAR        20.4          32            117   
9          qTn      BIGINT           0           7              2   
10           P     VARCHAR     733.933     838.454            469   
11          qP      BIGINT           0           7              2   
12           T     VARCHAR        20.5        33.0            148   
13          qT      BIGINT           0           7              2   
14           U     VARCHAR         100          99             48   
15          qU      BIGINT           0           7              2   
16          Ed     VARCHAR        14.2          29            164   
17         qEd      BIGINT           0           7              2   
18          Tw     VARCHAR          19          31            116   
19         qTw      BIGINT           0           7              2   
20           D     VARCHAR           0          90             16   
21          qD      BIGINT           0           7              2   
22           F     VARCHAR           0         5.6             55   
23          qF      BIGINT           0           7              2   
24           N     VARCHAR           0           9             10   
25          qN      BIGINT           0           7              2   
26          Px      DOUBLE       753.0       760.5             14   
27         qPx      BIGINT           0           0              1   
28          Pn      DOUBLE       751.5       759.0             16   
29         qPn      BIGINT           0           0              1   

                   avg                 std                q25  \
0    38.48485170617625   6.265772328680998                 32   
1                 None                None         1932-09-13   
2    1.076857659189965  0.7789338343146792                  1   
3                 None                None               None   
4   1.5087169129371745   2.878369651256062                  0   
5    13.47675530066041   5.540707349744609                8.0   
6                 None                None               None   
7    6.177973129872284  2.2536373088753066                  7   
8                 None                None               None   
9    6.177973129872284  2.2536373088753066                  7   
10                None                None               None   
11  2.6695560624670995  3.4001130244148006                  0   
12                None                None               None   
13  2.7746578366445918  3.4240905915955713                  0   
14                None                None               None   
15   5.272750598699965  3.0179685751315017                  5   
16                None                None               None   
17   5.272151898734177    3.01832020087677                  5   
18                None                None               None   
19  5.1188847074923025   3.103229921713479                  0   
20                None                None               None   
21   3.664043790626069   3.496303068601659                  0   
22                None                None               None   
23   3.837068080738967  3.4838804809228456                  0   
24                None                None               None   
25   5.857680465275402  2.5868736292233128                  7   
26   757.0765027322404  1.2201032108793177  756.3888888888889   
27                 0.0                 0.0 

#### Summarized

In [199]:
%%sql 
df << FROM (SUMMARIZE Aruba)

In [200]:
import numpy as np
print(np.average(df['null_percentage']))
print(np.average(df['count']))

44.07620689655171
17537.0


In [201]:
%%sql 
df << FROM (SUMMARIZE Bonaire)

In [202]:
print(np.average(df['null_percentage']))
print(np.average(df['count']))

55.48809523809524
20091.0


In [217]:
%%sql 
df << FROM (SUMMARIZE Curacao);

In [219]:
print(np.average(df['null_percentage']))
print(np.average(df['count']))
df.head(50)

77.88148936170212
62419.0


column_name column_type         min         max  approx_unique  \
0      column0      BIGINT           0         365            385   
1          Stn      BIGINT       78989       78989              1   
2        Datum        DATE  1898-01-01  1919-12-31           7321   
3       Niveau      BIGINT           1           1              1   
4           Rd     VARCHAR           0           T            119   
5          qRd      BIGINT           0           7              2   
6    column0_1      BIGINT           0        8759           7828   
7         Tijd      DOUBLE         1.0       800.0             32   
8            F     VARCHAR           0         9.8             41   
9           qF      BIGINT           0           7              3   
10    column00      BIGINT           0         365            385   
11          Tp     VARCHAR        23.9        31.4             32   
12         qTp      BIGINT           0           7              2   
13          Pt     VARCHAR      757.53       766.8            225   
14         qPt      BIGINT           0           7              2   
15          Pc     VARCHAR        6.27        7.04             30   
16         qPc      BIGINT           0           7              2   
17          P0     VARCHAR      747.04      761.21            267   
18         qP0      BIGINT           0           7              2   
19           P     VARCHAR       753.1       763.8             38   
20          qP      BIGINT           0           7              2   
21           T     VARCHAR        None        None              0   
22          qT      BIGINT           7           7              1   
23          Tw     VARCHAR        None        None              0   
24         qTw      BIGINT           7           7              1   
25          Td     VARCHAR       -12.7       -13.1              5   
26         qTd      BIGINT           0           7              2   
27           U     VARCHAR          73          81             10   
28          qU      BIGINT           0           7              2   
29          Ed     VARCHAR        18.9        20.2             10   
30         qEd      BIGINT           0           7              2   
31          Tx     VARCHAR        25.2        36.2             87   
32         qTx      BIGINT           0           7              2   
33          Tn     VARCHAR         1.9        36.1            108   
34         qTn      BIGINT           0           7              2   
35           D     VARCHAR           0          90             10   
36          qD      BIGINT           0           7              2   
37          Fx     VARCHAR        None        None              0   
38         qFx      BIGINT           7           7              1   
39         DFx     VARCHAR        None        None              0   
40        qDFx      BIGINT           7           7              1   
41           N     VARCHAR        None        None              0   
42          qN      BIGINT           7           7              1   
43           C     VARCHAR        None        None              0   
44          qC      BIGINT           7           7              1   
45          Dn     VARCHAR           0          90              5   
46         qDn      BIGINT           0           7              2   

                   avg                 std                q25  \
0   182.02673581942364  105.38346926636754                 91   
1              78989.0                 0.0              78989   
2                 None                None         1899-10-13   
3                  1.0                 0.0                  1   
4                 None                None               None   
5    4.871288213594361   3.220240269106866                  0   
6               4379.5  2528.8182191779947               2191   
7   126.70981592143418   265.6118209818324  7.771365407785331   
8                 None                None               None   
9   2.2830548390714367   3.2812384057053

In [205]:
%%sql 
df << FROM (SUMMARIZE Saba)

In [206]:
print(np.average(df['null_percentage']))
print(np.average(df['count']))

41.98374999999999
38726.0


In [207]:
%%sql 
df << FROM (SUMMARIZE Statia)

In [208]:
print(np.average(df['null_percentage']))
print(np.average(df['count']))

6.05
4383.0


In [209]:
%%sql 
df << FROM (SUMMARIZE Suriname)

In [210]:
print(np.average(df['null_percentage']))
print(np.average(df['count']))

21.586060606060606
25275.0


In [211]:
%%sql 
df << FROM (SUMMARIZE Willemstad)

In [212]:
print(np.average(df['null_percentage']))
print(np.average(df['count']))

39.8865
58803.0


Aruba & 17537 & 44.1\
Bonaire & 20091 & 55.5\
Curacao & 62419 & 77.9\
Saba & 38726 & 42.0\
St. Eustatius & 4383 & 6.1\
Suriname & 25275 & 21.6\
Willemstad & 58803 & 39.9